core/cache/application.go

```diff
@@ -28,18 +33,78 @@ type Application struct {
 	configHash string
 }
 
-func (m *Application) setDetails(details ApplicationChange) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	m.details = details
+func (a *Application) WatchFields(comparitors ...func(*ApplicationDelta) bool) *applicationFieldWatcher {
+	w := newApplicationFieldWatcher(comparitors)
+
+	unsub := a.hub.Subscribe(a.topic(applicationFieldChange), w.detailsChange)
+
+	w.tomb.Go(func() error {
+		<-w.tomb.Dying()
+		unsub()
+		return nil
+	})
+
+	return w
+}
+
+// TODO (manadart 2018-03-13) Should we change this up and down the call stack
+// so that pointers are passed instead of copies?
+func (a *Application) setDetails(details ApplicationChange) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	delta := &ApplicationDelta{old: a.details, new: details}
+	a.details = details
 
 	configHash, err := hash(details.Config)
 	if err != nil {
 		logger.Errorf("invariant error - application config should be yaml serializable and hashable, %v", err)
 		configHash = ""
 	}
-	if configHash != m.configHash {
-		m.configHash = configHash
+	if configHash != a.configHash {
+		a.configHash = configHash
 		// TODO: publish config change...
 	}
+
+	a.hub.Publish(a.topic(applicationFieldChange), delta)
+}
+
+// topic prefixes the input string with the application name.
+func (a *Application) topic(suffix string) string {
+	return a.details.Name + ":" + suffix
+}
+
+type applicationFieldWatcher struct {
+	*notifyWatcherBase
+
+	comparitors []func(*ApplicationDelta) bool
+}
+
+func newApplicationFieldWatcher(comparitors []func(*ApplicationDelta) bool) *applicationFieldWatcher {
+	return &applicationFieldWatcher{
+		notifyWatcherBase: newNotifyWatcherBase(),
+
+		comparitors: comparitors,
+	}
+}
+
+func (w *applicationFieldWatcher) detailsChange(topic string, value interface{}) {
+	delta, ok := value.(*ApplicationDelta)
```

Comment:
By passing this reference that contains both the new and the old, we avoid keeping a copy in each watcher.

Reply:
To point out a trade-off, doesn't this mean that the Watchers *have* to process every change, and there is no change for collapsing multiple changes into a single event? Since the watcher isn't comparing "current" state to "last known" state that *it* knew about.
And if we are losing that property of "evaluate current vs the last thing you evaluated", it seems much more efficient to get the list of *only what has changed* if they have to stay up-to-date all the time anyway.